src/main/java/org/cryptomator/cloudaccess/webdav/CachedPropfindEntryProvider.java

```diff
@@ -0,0 +1,230 @@
+			var data = new PropfindEntryItemData.Builder()
+					.withPath(path.toString())
+					.withCollection(false)
+					.withSize(Optional.of(size))
+					.withLastModified(lastModified)
+					.withEtag(eTag.get())
+					.build();
+			cache.getOrCreateCachedNode(path).update(data);
+			cache.markDirty(path.getParent());
+		} else {
+			cache.markDirty(path);
+		}
+	}
+
+	public void createFolder(CloudPath path) {
+		cache.markDirty(path);
+	}
+
+	public void delete(CloudPath path) {
+		cache.delete(path);
+	}
+
+	private class RemoteChangeDetector extends TimerTask {
+
+		@Override
+		public void run() {
+			LOG.trace("run executed");
+
+			try {
+				var rootPath = CloudPath.of("/");
+				var root = cache.getCachedNode(rootPath);
+				if (root.isPresent()) {
+					var rootItemData = rootPoller.apply(rootPath);
+					var localData = root.get().getData(PropfindEntryItemData.class);
+					if (localData == null || !rootItemData.isSameVersion(localData)) {
+						root.get().update(rootItemData);
+						updateChildren(rootPath);
+					}
+				}
+			} catch (Exception e) {
+				LOG.error("Problem occoured during RemoteChangeDetector running", e);
+			}
+		}
+
+		private void updateChildren(CloudPath node) {
+			LOG.trace("updateChildren {}", node);
+			var cacheNode = cache.getCachedNode(node);
+			if (cacheNode.isPresent()) {
+				var localChildren = cacheNode.get().getChildren();
+				var remoteChildren = cacheUpdater.apply(node);
+
+				// delete parent
+				remoteChildren.sort(new PropfindEntryItemData.AscendingByDepthComparator());
+				if(remoteChildren.size() > 0) {
+					remoteChildren.remove(0);
+				}
+
+				// if local exists but not remote --> remove including descendant ✓
+				// if local exists but different or no ETAG --> update, check descendant up to unchaged or not exist in cache ✓
+				// if local exists and same ETAG --> ignore further sub-tree ✓
+				// if remote exists but not local --> create ✓
+
+				for (CachedNode localChild : localChildren) {
+					var remoteItemMetadata = remoteChildren.stream().filter(remote -> remote.getName().equals(localChild.getName())).findAny();
+					if (remoteItemMetadata.isPresent()) {
+						var localData = localChild.getData(PropfindEntryItemData.class);
+						if (localData != null) {
+							if (!localData.isSameVersion(remoteItemMetadata.get())) {
+								cache.getCachedNode(node.resolve(localChild.getName())).get().update(remoteItemMetadata.get());
+								if (localData.isCollection()) {
+									updateChildren(node.resolve(localChild.getName()));
+								}
+							} // else branch does nothing, even if a subtree exists but ETag of the parent didn't change
+						} else {
+							cache.getCachedNode(node.resolve(localChild.getName())).get().update(remoteItemMetadata.get());
+							if(remoteItemMetadata.get().isCollection()) {
+								updateChildren(node.resolve(localChild.getName()));
+							}
+						}
+					} else {
+						cache.delete(node.resolve(localChild.getName()));
+					}
+				}
```

Comment:
high complexity, could use some private subfunctions

Reply:
Still not good but with https://github.com/cryptomator/cloud-access-java/pull/5/commits/bba82105a174f910fa8d848ea36978fadfd3d69b maybe a little bit better

Reply:
sufficient for now 👍